### PR TITLE
Deprecate ref-maybe-const for `forall` intents

### LINKED ISSUE #21398: are we sure about ref-maybe-const for 2.0?
We are preparing for the Chapel 2.0 release where the language is stable and frozen. This issue asks about whether or not we are confident that we want to keep the ref-maybe-const feature for 2.0.

### Background & History

The [abstract intents table](https://chapel-lang.org/docs/language/spec/procedures.html#abstract-intents-table) describes the current situation. The default intent varies between `const in` / `const ref` / `ref` depending on the type and other details. In particular:
 * for `atomic`, `sync`, and `single`, the default intent is `ref`
 * for arrays, the default intent is either `ref` or `const ref`. The compiler chooses `ref` if the body of the function modifies the array formal.
 * for the `this` formal for a method on a record, the default intent is similarly either `ref` or `const ref` & compiler chooses `ref` if the body of the function modifies the formal.

This design has evolved over time.

Historically, arrays were in the same category as `sync` / `single` / `atomic` and the default intent was `ref`. In 2012, we had some discussion [on an email thread](https://sourceforge.net/p/chapel/mailman/message/28928289/)) about potentially making these cases all `const ref`. Then, in 2017, we ran into some unintended consequences when combining default intent functions (such as `writeln`) with array element access that uses return intent overload (see issue #5217). In response to those issues, we considered 3 proposals: 1) just `const ref` #5218 2) ref-if-modified aka ref-maybe-const #5219 3) specialize-blank #5220 and went with ref-maybe-const. One reason for that is that the just-const-ref proposal ran into some challenges with combining with task intents and promotion.

Later, we have heard some feedback from a user that the default intents was too complicated and we sought to improve the documentation (PR #20700).

One more note -- partly because the default intent for arrays was historically `ref`, the compiler allows mutation of array call-expr temporaries (such as with `modifyArray(returnArray())`) even though it does not allow such mutation of literals/rvalues in other contexts (see [this comment on issue #17195](https://github.com/chapel-lang/chapel/issues/17195#issuecomment-826816907)).

We have also identified these related issues:
 * Should virtual methods disable ref-if-modified? #6497 
 * Syntax for ref-maybe-const intent #14822
 * CG: how to interpret abstract intents? #16988 
 
### New Problems Identified

This issue discusses a few new problems with ref-maybe-const that, for the most part, have not been brought up previously.

#### Compiler Architecture

Presently, we are both trying to stabilize the language as well as re-implementing the resolution piece as part of the *dyno* effort. The current ref-maybe-const design is difficult to implement in full generality without creating a sort of full-program resolution phase (which is the normal mode of operation for the production resolver, but not for the prototype dyno resolver). In the context of the dyno resolver, these problems are particularly apparent with recursive functions.

In the past, we have considered whether or not ref-maybe-const is problematic for separate compilation. I think it is manageable for separate compilation, but at the very least it does add some complexity.

#### Documentation and API Stability

When writing and publishing a library, it's important to be able to clearly document the API and to be able to offer API stability by being aware of cases where the API changes. The ref-maybe-const intent causes problems for both of these, because, for function signatures using default intent, the function signature alone isn't enough to tell you if you can pass a `const` variable as an argument. In particular, if the function accepts an array argument, then it depends on the body of the function whether the array formal is `const ref` or `ref`. But, the body of the function is not available in the documentation, and furthermore, somebody trying to understand how to they can use the function shouldn't have to look at the body of the function.

It's certainly the case that we could try to improve the tooling in this area. The `chpldoc` output could include results of resolution & indicate the concrete intents. However, that isn't always possible for a generic function, since the intent can vary depending on the type it is instantiated with. Additionally, even if the `chpldoc` output did clarify the situation, we would still have the problem that somebody creating a library could easily change one of these intents from `const ref` to `ref` in the process of updating the implementation. They might update the implementation and produce a new minor version without ever realizing that they are changing the API.

#### Interfaces

This one isn't really a new problem since it has its own issue #16988, but it deserves a few more words.

When writing an `interface`, we need to be able to write a function signature without a body. So, the ref-maybe-const rule cannot apply because it's not possible to analyze the function body. See also #16988 which asks what array-typed formals should mean in a routine in an interface.

#### Universal Method Call Syntax

One issue we face is how to request the version of a method for a particular interface. If resolved, this strategy can enable interfaces as a solution for special method naming. The way Rust handles it is to use universal method call syntax. With universal method call syntax, `receiver.method(arg)` and `method(receiver, arg)` are interchangeable. That means that a specific interface can be requested with `SomeInterface.method(receiver, arg)`. See also 
More https://github.com/chapel-lang/chapel/issues/19050#issuecomment-1375836834 . Note that we have not discussed unified method call syntax much yet (it probably deserves its own issue).

The point here is that if we adopted unified method call syntax, it would seem really weird to have the 1st argument to certain routines have a different default intent from the 2nd argument. It depends on the details of the universal method call syntax how bad this would be. (Can you call a function that isn't declared as a method as a method, like `a.f(b)` for `proc f(a, b)` ? If so, does the special receiver intent apply to `a`?)

### Possible Approaches

I'll add links to issues for possible approaches here as I write them.

### PR DESCRIPTION
**Pull request by jabraham17**
Deprecates ref-maybe-const for `forall` intents. The compiler will warn when an array is used in a modifiable context inside of a forall loop. For example passing the array to a function which takes a `ref` argument or assigning the array to a `ref` var.

Implements the rest of the decisions for https://github.com/chapel-lang/chapel/issues/21398#issuecomment-1654660481

Followup to #22958, #23049, and #23060

## Summary of changes
- enable warning for forall intents
- add an explicit `ref` where needed in modules and tests on foralls
- for tests in `test/optimizations/autoLocalAccess` and `test/optimizations/autoAggregation`, added a PREDIFF pending a fix for the optimizations tested here
  - using an explicit `ref` shadow variable for these causes the optimization to fail, this is a temporary solution until the optimization is fixed.
- enable `test/deprecated/ref-maybe-const-forall-intents.chpl`
- futurize `test/gpu/native/distArray/blockInsideOn.chpl`
  - using a forall with a shadow var for a distributed array inside of an `on gpu` block makes the loop gpu ineligible

## Testing
- [x] paratest with futures
- [x] paratest with gasnet
- [x] gpu testing
  - [x] nvidia with UM
  - [x] amd with AOD
- [x] paratest with c backend on nightly test machine

[Reviewed by @vasslitvinov]

closes #21398
closes https://github.com/Cray/chapel-private/issues/5093

### CODE REVIEW

File: compiler/resolution/lowerForalls.cpp [hunk 1]
```diff
@@ -887,9 +887,8 @@ static void expandTaskFn(ExpandVisitor* EV, CallExpr* callToTFn, FnSymbol* taskF
   if (addErrorArgToCall)
     addDummyErrorArgumentToCall(callToTFn);
 
-  // If we don't flatten it right away, we get non-global taskFns
-  // in expandTaskFn(). That may cause issues with scoping.
   flattenNestedFunction(cloneTaskFn);
+
```
Comment (vasslitvinov) [inline compiler/resolution/lowerForalls.cpp:891]: What is the reason for removing the comment?

The addition of an empty line seems unwarranted.

Comment (jabraham17) [reply compiler/resolution/lowerForalls.cpp:891]: This is a mistake, I will revert it. Thank you!

File: test/studies/bale/histogram/histo-atomics.chpl [hunk 2]
```diff
@@ -46,16 +46,16 @@ proc main() {
...
     when Mode.ordered {
-      forall r in rindex do
+      forall r in rindex with (ref A) do
         A[r].add(1);
     }
     when Mode.unordered {
-      forall r in rindex do
-	A[r].unorderedAdd(1);
+      forall r in rindex with (ref A) do
+	      A[r].unorderedAdd(1);
```
[the remark this answers is not in the document]
Comment (jabraham17) [reply test/studies/bale/histogram/histo-atomics.chpl:54]: These are spaces. I don't know why the diff renders them as tabs.

**Comment by bradcray** [comment]
Nice!!!

**Pull request merged**